refactor(conditions): Remove unused condition erase functionality

petervdmeer opened the pull request:
**Refactor**

This PR removes complex condition-erase functionality that was never used

## Summary
ConditionsStore had an Erase function, but no code in ES  explicitly erases conditions. Erasing an condition is done implicitly by setting the condition to zero. (And conditions set to zero are never written out to savegames, thus effectively erasing them at save.)

Removing this unused functionality should make the codebase simpler.

## Usage examples
N/A

## Testing Done
Removed the affected functions and compiled the code. The code was never called.

## Wiki Update
N/A

## Performance Impact
This should have no impact. The functions were never used.

tibetiroka reviewed (approved):
Looks good. I'm on mobile so I can't test it, but the changes seem straightforward.

warp-core commented on source/ConditionsStore.h:
```diff
@@ -114,11 +112,10 @@ class ConditionsStore {
 
-	// Add a value to a condition, set a value for a condition or erase a
-	// condition completely. Returns true on success, false on failure.
+	// Add a value to a condition or set a value for a condition. Returns true on
+	// success, false on failure.
```
Can we leave the second sentence on its own line so the break isn't mid-sentence?
```suggestion
	// Add a value to a condition or set a value for a condition.
	// Returns true on success, false on failure.
```

warp-core commented on tests/unit/src/test_conditionsStore.cpp:
```diff
@@ -235,18 +213,16 @@ SCENARIO( "Setting and erasing conditions", "[ConditionStore][ConditionSetting]"
 				REQUIRE( store["myFirstVar"] == 10 );
 			}
-			THEN( "erasing the condition will make it disappear again" )
+			THEN( "erasing the condition is done by setting it to zero" )
```
Setting it to zero and erasing aren't the same thing at runtime.
The entry still exists if it is set to zero, it just returns a value of zero, the same as if it were erased.
In terms of containers, "erase" generally means for the entry to no longer be present.

Assuming we already have tests for non-existent or zero values returning false, I'm not sure this test is necessary anymore. It's checking that, upon erasure, the relevant entry is removed from the `storage` container. Since that functionality no longer exists, and setting a condition to zero should not result in that entry being immediately removed from the `storage` container, there's not much point in this test anymore.

petervdmeer replied:
Agreed, I changed the tests to explicit set tests.

warp-core commented on tests/unit/src/test_conditionsStore.cpp:
```diff
@@ -415,7 +391,7 @@ SCENARIO( "Providing derived conditions", "[ConditionStore][DerivedConditions]"
 			THEN( "readonly providers should not perform erase actions" )
 			{
 				mockProvNamed.SetRONamedProvider(store, "named1");
-				REQUIRE_FALSE( store.Erase("named1") );
+				REQUIRE_FALSE( store.Set("named1", 0) );
```
This should probably be changed to testing that read-only providers reject "Set"ting their values.

petervdmeer replied:
Agreed, I updated the title of the test.

warp-core commented on tests/unit/src/test_conditionsStore.cpp:
```diff
@@ -468,7 +444,7 @@ SCENARIO( "Providing derived conditions", "[ConditionStore][DerivedConditions]"
 			THEN( "read-only prefixed provider should reject erase" )
 			{
 				mockProvPrefixA.SetROPrefixProvider(store, "prefixA: ");
-				REQUIRE_FALSE( store.Erase("prefixA: test") );
+				REQUIRE_FALSE( store.Set("prefixA: test", 0) );
```
As above, this is testing "Set" functionality, not erasure.

petervdmeer replied:
Agreed, I updated the description to talk about setting.

Zitchas commented:
Based on failing checks on Delta, this PR may have removed functionality that https://github.com/endless-sky/endless-sky/pull/8082 made use of.

warp-core commented:
> Based on failing checks on Delta, this PR may have removed functionality that https://github.com/endless-sky/endless-sky/pull/8082 made use of.

It looks like that PR is calling `ConditionsStore::SetEraseFunction` in Player info (the site of the compile error) which is removed by this PR because the erase function it sets is never used. It doesn't appear that that PR is changing that. So removing the expression making the call to the removed function will fix the compile errors.

Zitchas commented:
I'm sorry, but I'm having trouble parsing what you said. 

Are you saying that #8082 doesn't actually use the function, just calls it without purpose?

Or are you making the observation that 8082 just needs to figure out an alternative to using this?

warp-core commented:
> Are you saying that #8082 doesn't actually use the function, just calls it without purpose?

I believe this is the most accurate of your descriptions.

petervdmeer commented:
> Are you saying that #8082 doesn't actually use the function, just calls it without purpose?

The most proper statement is that I added some template calls without a purpose a few years ago for automatic conditions, and I corrected that in this PR. The author of #8082 correctly followed the template, but the template is now changed because I corrected my earlier mistake.

> Or are you making the observation that 8082 just needs to figure out an alternative to using this?

The author of #8082 just needs to see that one of the two calls is no longer needed and remove it from the code. The extra call was actually never needed, but I only realized that recently.
